From ed569ba9ca55f6caebd33801a710679abf20b35b Mon Sep 17 00:00:00 2001
From: Kishor PK <kpbhat@codeaurora.org>
Date: Wed, 24 May 2017 15:10:03 +0530
Subject: [PATCH 950/959] soc: qcom: pil: Avoid possible buffer overflow during
 Modem boot

Buffer overflow can occur if MBA firmware size exceeds 1MB.
So validate size before copying the firmware.

Mot-CRs-fixed: IKSIMP-5347
CVE-Fixed:CVE-2017-8243
CRs-Fixed: 2001803

Change-Id: I070ddf85fbc47df072e7258369272366262ebf46
Signed-off-by: Kishor PK <kpbhat@codeaurora.org>
Signed-off-by: Amarendra Reddy <amarenr@motorola.com>
Reviewed-on: https://gerrit.mot.com/1004675
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Igor Kovalenko <igork@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/soc/qcom/pil-msa.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/qcom/pil-msa.c b/drivers/soc/qcom/pil-msa.c
index 701edbca0b2..77577ac4919 100644
--- a/drivers/soc/qcom/pil-msa.c
+++ b/drivers/soc/qcom/pil-msa.c
@@ -562,7 +562,16 @@ int pil_mss_reset_load_mba(struct pil_desc *pil)
 		goto err_mba_data;
 	}
 	count = fw->size;
-	memcpy(mba_virt, data, count);
+
+	if (count <= SZ_1M) {
+		/* Ensures memcpy is done for max 1MB fw size */
+		memcpy(mba_virt, data, count);
+	} else {
+		dev_err(pil->dev, "%s fw image loading into memory is failed due to fw size overflow\n",
+			__func__);
+		ret = -EINVAL;
+		goto err_mss_reset;
+	}
 	wmb();
 
 	/* Load modem debug policy */
-- 
2.11.0

